### PR TITLE
Allows Ian to eat homework

### PR DESCRIPTION
**Pull request by SabreML**
## What Does This PR Do

Adds a small (10%) chance for Ian to eat any paperwork that's lying around on the floor near him.
This can of course be avoided by taking the paper away from Ian, but if he is allowed to "play" with the it for too long then one of two things can happen:
* **90%** chance of the paper being crumpled up, and any text on it being made illegible.
* **10%** chance of Ian simply swallowing the paper whole.

This also works for player-controlled Ians, so any prospective "Paperwork HoPs" had better get into the habit of using a clipboard or filing cabinet to store their vital documents.

*(Also refactored the Ian AI code a little.)*

## Why It's Good For The Game

I'd written up a long and in-depth justification for why this PR should be merged, but sadly my dog ate it a few minutes ago.
I guess you'll just have to take my word for it!

## Images of changes

**The tasty looking paperwork:**
before

**After a bit of chewing:**
paper

## Changelog
:cl:
add: Ian now has a small chance to eat any important paperwork on the floor near him.
tweak: Increased the chance of Ian begging for food in your hand by 10%.
/:cl:

### CODE REVIEW

**Comment by SabreML** [comment]
I have been informed that due to British Summer Time, I am ahead by one hour.
Please accept my most sincere apologies for this error.

**Comment by AffectedArc07** [comment]
This honestly isnt even that bad of a PR, like you could have opened this any other day and it would have a high merge chance.

**Comment by SabreML** [comment]
> This honestly isnt even that bad of a PR, like you could have opened this any other day and it would have a high merge chance.

That's what I was aiming for really. No reason not to get it exactly on the mark UTC-wise though.

File: code/modules/mob/living/simple_animal/friendly/dog.dm [hunk 1]
```diff
@@ -503,53 +503,63 @@
...
+
+/mob/living/simple_animal/pet/dog/corgi/Ian/proc/move_to_target()
+	stop_automated_movement = TRUE
+	step_to(src, movement_target, 1)
+	sleep(3)
+	step_to(src, movement_target, 1)
+	sleep(3)
+	step_to(src, movement_target, 1)
+
+	if(movement_target) // Not redundant due to sleeps, Item can be gone in 6 decisecomds
```
Comment (Adrer) [inline code/modules/mob/living/simple_animal/friendly/dog.dm:545]: Minor nitpick but comment say decisecomd rather than decisecond and it's annoying me.

**Comment by LightFire53** [comment]
I vote we keep it.

**Comment by warior4356** [comment]
I'm also in favor of this being made slightly more serious as an actual PR, as this seems like a good addition to the game.

**Comment by meow20** [comment]
I thought it means the sloth because first april and all...

**Comment by warriorstar-orion** [comment]
YES PLEASE

**Comment by Dumbdumn5** [comment]
I'm absolutely game for this, but I'd say, rather than being chance based, is it possible to make this an action with a cooldown? I love the 'Ian ate my paperwork' joke, but if they get a fax from CC then it's kinda bad to just have that chomped due to an AI controlled dog walking near it before the dude can get there. Having it be an intentional thing that player controlled Ians and corgis can do would be hilarious though.

**Review by S34NW — APPROVED** [review]
Chat messages are good for readability and grammar!

**Review by Kyep — APPROVED** [review]
I'd prefer it if 'eaten' paper had a different sprite, or just displayed:
"This paper has been chewed so badly it is unreadable"
instead of random characters.

**Comment by SabreML** [comment]
> I'm absolutely game for this, but I'd say, rather than being chance based, is it possible to make this an action with a cooldown? I love the 'Ian ate my paperwork' joke, but if they get a fax from CC then it's kinda bad to just have that chomped due to an AI controlled dog walking near it before the dude can get there. Having it be an intentional thing that player controlled Ians and corgis can do would be hilarious though.

I don't think the AI behaviour will be an issue for any faxes, primarily because it ignores any paper which is on a table (Which fax machines are also on).
Player controlled Ians could be an issue here since they don't have any such restrictions, and can just click on a paper to start eating it. Honestly not sure how to solve that one, maybe by adding a longer cooldown or windup if it's a player doing it? Either way 99% of incidents will be from a player rather than the AI, and there is still a time period where he can be stopped if someone notices.

<hr>

> I'd prefer it if 'eaten' paper had a different sprite, or just displayed:
> "This paper has been chewed so badly it is unreadable"
> instead of random characters.

It does have a sprite as well as the text changing, if that's what you mean.
scrap
I had considered just putting something like that instead of making the existing text gibberish, but I thought it would be more interesting if you could *just* barely make out what it said originally.

File: code/modules/paperwork/paper.dm [hunk 2]
```diff
@@ -163,6 +163,50 @@
...
+
+	// 90% chance of a crumpled paper with illegible text.
+	if(prob(90))
+		var/message_ending = "."
+		var/obj/item/paper/crumpled/P = new(loc)
+		P.name = name
+		if(info) // Something written on the paper.
+			var/new_text = strip_html_properly(info, MAX_PAPER_MESSAGE_LEN, TRUE) // Don't want HTML stuff getting gibberished.
+			P.info = Gibberish(new_text, 100)
+			message_ending = ", the drool making it an unreadable mess!"
```
Comment (AffectedArc07) [inline code/modules/paperwork/paper.dm:195]: Put a hardcoded string here for the paper being unreadable without the text being turned into gibberish. If someone eats paper with a long enough message on it, it will crash the server. This is why paper being burnt doesn't turn the message to gibberish.

**Review by AffectedArc07 — CHANGES_REQUESTED** [review]

File: code/modules/paperwork/paper.dm (hunk 2 above)
Comment (SabreML) [reply code/modules/paperwork/paper.dm:195]: It... how?
The `strip_html_properly()` proc is capping the length to 3072 characters maximum, and the `Gibberish()` is coming after that.
I tested it after running this code and it all worked fine, not even any lag spikes. (On a private test server)
```dm
/obj/item/paper/proc/aaaa()
	var/list/test = list()
	for(var/I in 1 to 100000)
		test += "[I]"
	info = jointext(test, "")
	return length(info) // 488895 characters
```
I'll remove it if it's an issue, but having tested it so much I can't see how it would be.

Comment (AffectedArc07) [reply code/modules/paperwork/paper.dm:195]: Its happened before and took us months to pin down because it doesnt use hrefs. Remove it.

File: code/modules/mob/living/simple_animal/friendly/dog.dm [hunk 3]
```diff
@@ -503,53 +503,63 @@
...
+			dir = NORTH
+		else // Face south if you're on or above the tile
+			dir = SOUTH
+
+		if(!Adjacent(movement_target)) //can't reach food through windows.
+			return
 
-		if(prob(1))
-			custom_emote(1, pick("dances around.","chases its tail!"))
-			spin(20, 1)
```
Comment (dearmochi) [inline code/modules/mob/living/simple_animal/friendly/dog.dm:552]: Why was this removed?

File: code/modules/mob/living/simple_animal/friendly/dog.dm [hunk 4]
```diff
@@ -503,53 +503,63 @@
...
+	if(movement_target) // Not redundant due to sleeps, Item can be gone in 6 deciseconds
+		// Face towards the thing
+		if(movement_target.loc.x < src.x)
+			dir = WEST
+		else if(movement_target.loc.x > src.x)
+			dir = EAST
+		else if(movement_target.loc.y > src.y)
+			dir = NORTH
+		else // Face south if you're on or above the tile
+			dir = SOUTH
```
Comment (dearmochi) [inline code/modules/mob/living/simple_animal/friendly/dog.dm:554]: Could probably just use [`get_dir`](https://secure.byond.com/docs/ref/index.html#/proc/get_dir)

File: code/modules/mob/living/simple_animal/friendly/dog.dm [hunk 5]
```diff
@@ -503,53 +503,63 @@
...
+
+	if(++turns_since_scan > 5)
+		turns_since_scan = 0
+
+		// Has a target, but it's not where it was before, and it wasn't picked up by someone.
+		if(movement_target && !(isturf(movement_target.loc) || ishuman(movement_target.loc)))
+			movement_target = null
+			stop_automated_movement = FALSE
+		// No current target, or current target is out of range.
+		if(!movement_target || !(movement_target.loc in oview(src, 3)))
```
Comment (dearmochi) [inline code/modules/mob/living/simple_animal/friendly/dog.dm:517]: `oview(src, 3)` could probably be cached right before the `if` so it's not recomputed again the following loop

**Review by dearmochi — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/simple_animal/friendly/dog.dm (hunk 3 above)
Comment (SabreML) [reply code/modules/mob/living/simple_animal/friendly/dog.dm:552]: It was replaced with
```dm
	if(prob(1))
		chasetail()
```
Since that verb does more or less exactly the same thing.
```dm
/mob/living/simple_animal/pet/dog/verb/chasetail()
	set name = "Chase your tail"
	set desc = "d'awwww."
	set category = "Dog"

	visible_message("[src] [pick("dances around", "chases [p_their()] tail")].", "[pick("You dance around", "You chase your tail")].")
	spin(20, 1)
```

Comment (dearmochi) [reply code/modules/mob/living/simple_animal/friendly/dog.dm:552]: But isn't the code I mentioned part of the dog's AI instead of being a player-facing verb?

Comment (SabreML) [reply code/modules/mob/living/simple_animal/friendly/dog.dm:552]: It is, yes. I changed the AI code to call the `chasetail()` verb rather than `custom_emote()` and `spin()`, since what was there was pretty much 1:1 with the verb.

**Review by dearmochi — APPROVED** [review]

**Review by AffectedArc07 — APPROVED** [review]
